File: Algorithm/Python/034. Search for a Range.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-04-08 15:02:26
# @Last modified by:   LC
# @Last Modified time: 2016-04-10 16:22:30
# @Email: liangchaowu5@gmail.com



# 找到与target相同的数，从这个位置往左右递归找
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if len(nums) == 0:
            return [-1,-1]
        
        posi = [-1,-1]
        left = 0
        right = len(nums)-1
        self.find(nums,target,left,right,posi)
        return posi
                
    def find(self,nums,target,left,right,position):
        while left < right:
            mid = (left + right)/2
            if nums[mid] < target:
                left = mid + 1
            elif nums[mid]> target:
                right = mid - 1
            else:
                if position[0] == -1 or position[0]>mid:
                    position[0] = mid
                if position[1] == -1 or position[1]<mid:
                    position[1] = mid
                self.find(nums,target,left,mid-1,position)
                self.find(nums,target,mid+1,right,position)
                return           # 这里需要有返回，否则程序回到上一次找到这个值的地方会一直死循环递归下去
        if nums[left]==target:
            if position[0] == -1 or position[0]>left:
                position[0] = left
            if position[1] == -1 or position[1]<right:
                position[1] = right
            
                
            
        

File: Algorithm/Java/034. Search for a Range.java
/**
* Author: LC
* Date:   2016-12-02 10:23:20
* Last modified by:   WuLC
* Last Modified time: 2016-12-02 10:25:49
* Email: liangchaowu5@gmail.com
*/


// binary search
// use two binary search, one to find the insert position of target, another to find the insert position of target+1
// both insertion should maintain the nums in order, when there are duplicates, insert at the first position
public class Solution 
{
    public int[] searchRange(int[] nums, int target) 
    {
        int[] result = {-1, -1};
        if (nums.length == 0 || target < nums[0] || target > nums[nums.length-1]) return result;
        int leftInsertPoint = insertPosition(nums, target);
        int rightInsertPoint = insertPosition(nums, target + 1);
        if (rightInsertPoint != leftInsertPoint) 
        {
            result[0] = leftInsertPoint;
            result[1] = rightInsertPoint-1;
        }
        return result;
    }
    
    public int insertPosition(int[] nums, int target) // the first insert position of target to keep nums in order
    {
        int left = 0, right = nums.length-1, mid;
        while (left < right)
        {
            mid = left + ((right-left)>>1);
            if (nums[mid] >= target) right = mid;
            else left = mid + 1;
        }
        // left = right
        if (nums[left] >= target) return left;
        else return left + 1;
    }
}